Don't cache content

niran opened the pull request:
Caching the content correctly (e.g. respecting Vary headers) on our own would require duplicating the CacheMiddleware. Instead, cache the ESI URLs included in a response based on the hash of the content. For uncached responses, search for ESI tags in the content and cache the URLs.

tswicegood committed:
Refactoring to use a list of URLs generated at the template tag level

This commit demonstrates the problem with testing behavior rather than
state.  My bad... :-/

tswicegood committed:
Bit of refactoring and stop explicitly relying on behavior of BaseEsiMiddleware

Attempted to get rid of BaseEsiMiddleware all together and move its
process_request() into ResponseMiddleware, but that causes issues with
the RequestMiddleware handling as expected.  Controlling the explicit
execution the way it does now is required in order to make sure that
cache hits return the response.

tswicegood committed:
add another task

tswicegood committed:
Add another warning...

niran committed:
Refactor ESI tag replacement into one common function. Untested.

niran committed:
Add request to replace_esi_tags function to pass to the view

niran committed:
Consolidate duplicate decorators and use a fake request decorator for all fake requests

niran committed:
Add a local HTTP client and use it for fetching ESI endpoints

niran committed:
Copy over django.test.client

niran committed:
Copy our customizations from http to client

niran committed:
Rename client to http_client

niran committed:
Switch from using http to http_client

niran committed:
Remove armstrong.esi.http

niran committed:
Remove testing code from http_client.Client.request

niran committed:
Remove login and logout from http_client.Client since they aren't useful for our needs

niran committed:
Remove unneeded libraries

niran committed:
Correct reference to client instance

niran committed:
Remove exception storage from Client

niran committed:
Remove resolver from ResponseMiddleware

niran committed:
Remove resolver-specific tests

niran committed:
Propagate exceptions during tests instead of trying to display the 500 page

niran committed:
Use a real view instead of a mocked view in a test

niran committed:
Replace another fake view in a test with a real view

niran committed:
Replace assertNotRegexpMatches with assertFalse on an re.search call to run tests using python 2.6

niran committed:
Remove tests that test old resolver logic

niran committed:
Update another test to use real views instead of the fake ones and the resolver

niran committed:
Add http_client tests

niran committed:
Only store the ESI endpoint URLs in the cache since we no longer need the views at this level

niran committed:
Add Django to test requirements

niran committed:
Update test to use the new cache format

niran committed:
Remove unused import from tests

niran committed:
Pass the current request URL as the referrer to ESI fragments

niran committed:
Merge appropriate headers and cookies from ESI fragments into the main request

niran committed:
Rename _esi_was_invoked to _esi_fragment_urls

niran committed:
Store and fetch ESI URLs from the cache based on the content of the response.

Caching the content correctly (e.g. respecting Vary headers) on our
own would require duplicating the CacheMiddleware. Instead, cache
the ESI URLs included in a response based on the hash of the content.
For uncached responses, search for ESI tags in the content and cache
the URLs.

niran committed:
Move replace_esi_tags to _utils

tswicegood commented:
See my comments above.  I'm uncomfortable with changes that require every response to be parsed on the thought that it might contain ESI values.  I wouldn't want to use a third-party caching tool that regex'd every page I generated.

niran committed:
Rename armstrong.esi._utils to armstrong.esi.utils

niran committed:
Replace ESI tags as we find them instead of searching the whole response for each tag.

This change had a few consequences:
 - Since we use the URLs as we find them, we no longer need to store
   them, so we now store whether ESIs were used instead.
 - Nothing was really being done in process_request, so the code
   that expects the ESI status to be on the request just uses a
   default value instead.
 - One of the tests no longer made sense, but it was replaced by
   a change to the context processor tests that tests the same
   logic.

niran committed:
Store the ESI status on the response in a separate middleware

This allows the ESI status to be cached with the response so we don't
have to cache it separately. It also means that we will always know
whether a response contains ESI tags without having to search the
content.

niran committed:
Verify that multiple ESI tags interspersed within content get replaced correctly

niran committed:
Remove unneeded and potentially detrimental signal sending from http_client

niran committed:
Add support for everything {% url %} can do to {% esi %}

niran committed:
Pass an X-ESI-Fragment HTTP header along to ESI fragment requests

tswicegood commented:
Merged in as of aac30ee2d1ce450e56d47e89ed837947551aa1d5 with some comments.
